kylin-spark-project/kylin-spark-common/src/main/scala/org/apache/spark/sql/execution/datasource/FilePruner.scala

```diff
@@ -170,7 +172,8 @@ class FilePruner(cubeInstance: CubeInstance,
   }
 
   private def genShardSpec(selected: Seq[SegmentDirectory]): Option[ShardSpec] = {
-    if (!KylinConfig.getInstanceFromEnv.isShardingJoinOptEnabled || selected.isEmpty) {
+    val isJoinQuery = OLAPContext.getThreadLocalContexts.map(_.allOlapJoins).exists(_.nonEmpty);
```

Comment:
shard join optimization is for joining two or more cubes, but not fact-table joining join-table.